### DISCUSSION TITLE
Stuck on debugging minified react hydration errors

### DISCUSSION THREAD

**Discussion by aleksakojadinovic** [vercel/next.js#59461] [Help] [answered]
### Summary

Hello,

First of all I'm well aware of the many existing discussion and issues on this topic such as https://github.com/vercel/next.js/issues/56441, https://github.com/vercel/next.js/issues/50245, https://github.com/vercel/next.js/issues/43159 and so on. Also I apologize in advance if this issue belongs in the react repository and not Next.js, but I assumed I'd find more help here, since Next.js is the most common framework where this error takes place.

Short summary of my situation: 
 - I'm using the pages router.
 - About 10 days ago I've updated from react 17 and next 12.3.2 to latest - that is react 18.2.0 and next 14.0.3.
 - Almost immediately I started receiving Minified react errors 418, 423 and 425.
 - These errors appear **only in production** and I have yet been **completely unable to reproduce them in development**.
 - These errors appear about twice a day, not too common but I cannot ignore them.

Now for the common answers:

 - I'm not server-side rendering any date-time data whatsoever. I'm very sure of this as I've spent the last 10 days eliminating all possible causes.
 - I'm not conditionally rendering anything based on whether the current environment is the client or the server (i.e. based on the window variable).
 - I'm sure to a high extent that it is not caused by invalid HTML nesting:
   - partially by inspecting components manually.
   - partially by the fact that no warnings appear in development builds, in all possible scenarios.
 - I have removed all dynamic imports, and the errors still appear.
 - I have tried deploying my application with source maps enabled (and thereby exposing some of my source-code for the purposes of debugging), but even the source-mapped errors are not helpful since react is still minified.
 - I have tried some hacks with disabling minification in production builds like shown [here](https://davidwalsh.name/how-to-not-minify-source-with-webpack) but to no avail, since these builds get broken immediately, partially because of the state of NODE_ENV variable, and some other very random errors. I though this path was unworthy of further exploration.
 - I have tried the hack presented in this issue: https://github.com/vercel/next.js/discussions/36641#discussioncomment-5215494 which is supposed to hack into the react-dom internals and provide more info about the error, namely the component stack trace, but this failed to work. The code was simply not executed when the error occurs (I have tested this by manually causing these errors locally by intentionally rendering different stuff on client and server).
 - I also have to point out that other than not being able to reproduce in development, **I'm also unable to reproduce this in production**, that is, the only info I have of the error is the logs I receive from actual users to whom this had happened, through my client-side logging system which is triggered on `window.onerror` and 
`window.onunhandledrejection`.
 - I also want to point out that the debugging method of removing parts of the page and seeing when the error will stop is  not acceptable in my case, since I have to wait 1 or maybe 2 whole days of no errors being reported in order to conclude that the error is removed, all the while affecting all other users who had not had the error in the first place.
 - React's `surpressHydrationWarnings` is usesless in my case since:
   - it only works 1 level deep.
   - I have no idea which component is causing these errors.

Now I am 10 days into debugging this and reading infinite issues and discussion, and it had lead me to this point where I am completely out of ideas. The worst part about this is the complete inability of reproduction both in development as well as in production, which makes my hands tied from the start. 

I cannot share any of my source code as it is private, and I cannot create a reproducible example because, as I've mentioned a lot, I cannot reproduce it myself.

So, does anyone have any idea how I should proceed about this? This is coming to a point where I might have to revert back to react 17 and next 12 simplify to mitigate any business impacts.

Thanks in advance :)

**Comment by icyJoseph** [comment]
Got any links to look at? Screenshots? Error logs?

The problem can be caused by Browser extensions which modify the DOM.

Got more info about the browsers that run into these issues?

Could it be locale issues? See here: https://github.com/vercel/next.js/issues/46038#issuecomment-1434841868

**Comment by aleksakojadinovic** [comment]
Hi, thanks for the quick reply, and sorry for not getting back to you earlier. However this is due to some good news - I've managed to identify where the problem is and have patched it to prevent impacting the users, but I still have no idea of the exact cause.

Namely the error occurs with, bear with me, **strings that have two or move consecutive spaces and are not wrapped in an element**. So here's how it went down: I fluked a registered user in my error logs and managed to retrieve the exact state of the page they had encountered when it crashed. I replicated the exact state and managed to reproduce it **but still only in production**.

Basically I first let the error crash and saw these logs:

<details>
<summary>
Screenshot of the client-side error log
</summary>
<img src="https://github.com/vercel/next.js/assets/49127698/de07209e-5b4e-4c16-9cd8-89a22833f5ac" />
</details>

Then I clicked on the first file in the trace and set a breakpoint like so: 
![image](https://github.com/vercel/next.js/assets/49127698/7453cf88-5bea-4a1e-b77e-fa0b3fc82a0a)

I reloaded the page and by inspecting the currently scoped variables (minified here as _e_ and _n_) I deduced that this was a mismatch between two strings where one of them has two whitespaces at a certain place and another has one:

![image](https://github.com/vercel/next.js/assets/49127698/f5acf364-a486-497a-8c7f-4f2378f2ba99)

This gave me some idea of what I'm dealing with. In order to minimize user impact I have simply sanitized these strings on our actual pages, and have created a few test pages to try and reproduce some other examples. After a few hours of deploying and testing, this is my conclusion:

**This causes an error**
```js
function testpage12() {
  return `String with  two spaces`;
}

export default testpage12;

export const getServerSideProps = () => ({ props: {} });
```

**This does NOT cause an error**
```js
function testpage12() {
  const str = `String with  two spaces`;
  return <span>{str}</span>;
}

export default testpage12;

export const getServerSideProps = () => ({ props: {} });
```

So as I've said in the beginning the error seems to only appear when rendering **string literals that have two or more consecutive whitespaces without a direct wrapper**. 

This is as far as I've managed to debug it and I do not know how to proceed honestly. Predicting where stuff like this will render is nearly impossible due to the fact that a lot of this happens in some external packages that we use, so we would have to sanitize every single string we use in advance. 

Few more remarks:
 - This still only happens in production, and what's even more peculiar is that it does not happen when I run a production build locally, only in actual production environment (we self-host on EC2, not Vercel)
 - I've seen something about whitespaces [here](https://react.dev/reference/react-dom/client/hydrateRoot#:~:text=use%20state%20instead.-,Pitfall,-The%20React%20tree) but I'm unsure whether this is a coincidence, does it have anything to do with whitespace in this context?

Any ideas what might be causing this, given this new info? Thanks :)

**Reply by icyJoseph** [reply]
Could you check somehow if React 17 also causes this? I mean, in React 17, it probably did happen, but the hydration error did not take down the page...

I wonder if this has to do with locales still. I'll use some time to investigate later today

**Reply by aleksakojadinovic** [reply]
> Could you check somehow if React 17 also causes this? I mean, in React 17, it probably did happen, but the hydration error did not take down the page...

Unfortunately I cannot verify it now since I would have to deploy the react 17 version to the actual users in order to test it, which we are unable to do right now.

> I wonder if this has to do with locales still.

Initially I though this was a locale issue, and I looked into all possibly localized values such as currencies and dates, but had not found anything. What locale-specific parameter do you think could impact whitespaces?

>  I'll use some time to investigate later today

Thanks for taking the time, your help is much appreciated.

**Reply by aleksakojadinovic** [reply]
Hi there, another update: Last night we had another production error, it was a logged in user so again we got lucky in that regard. This time it was again a double-spaced string, however now it was wrapped in an ```<h3>``` element, so apparently it might not be due to the string not having a direct HTML wrapper.

I cannot share the string itself due to data privacy, however the _e_ and _n_ values from the react minified code were of this form:
```
e: "The string that caused the crash "
n: "The string that caused the crash  "
```

**Reply by icyJoseph** [reply, marked as answer]
Is it content coming from a CMS? Maybe there's CRLF or other control characters, messing things up?

**Reply by aleksakojadinovic** [reply]
Yes, this is it, your comment got me thinking what's different between my local and production setup, and architecturally there's really only one difference - **Cloudflare**. So the issue was in fact caused by cloudflare minifying the HTML which includes whitespace removal, as shown [here](https://www.ryanmr.com/posts/cloudflare-minified-next).

I've disabled the minification for now and the problem disappeared. Thank you for the help.

I'll mark your comment as an answer, but while we are here, do you have any suggestions on long term solution to this issue? Cloudflare being one of the most common CMSs and HTML minification being a good practice, a few questions arise:
- Should Cloudflare expose an option not to trim whitespaces?
- Should React 18 hydration ignore whitespace differences in text content?
- Should it be the responsibility of the developer to always sanitize strings before rendering them?

I am aware that I am stepping out of scope of Next.js with these questions, but since this is a discussion and not an issue I hope it's fine.

Also a followup question if you don't mind - Do you think React should provide an easier way of debugging hydration errors 
in production, as many people are facing these issues, and even if you know who the culprit is it is very difficult to debug?

Thanks again for the help :)
